moveit_planners/trajopt/include/problem_description.h

```diff
@@ -0,0 +1,286 @@
+    TermInfoPtr out(new classname());                                                                                  \
+    return out;                                                                                                        \
+  }
+
+
+struct BasicInfo
+{
+  /** @brief If true first time step is fixed with a joint level constraint*/
+  bool start_fixed;
+  /** @brief Number of time steps (rows) in the optimization matrix */
+  int n_steps;
+  std::string manip;
+  std::string robot;             // optional
+  sco::IntVec dofs_fixed;             // optional
+  sco::ModelType convex_solver;  // which convex solver to use
+
+  /** @brief If true, the last column in the optimization matrix will be 1/dt */
+  bool use_time = false;
+  /** @brief The upper limit of 1/dt values allowed in the optimization*/
+  double dt_upper_lim = 1.0;
+  /** @brief The lower limit of 1/dt values allowed in the optimization*/
+  double dt_lower_lim = 1.0;
+};
+
+/**
+Initialization info read from json
+*/
+struct InitInfo
+{
+  /** @brief Methods of initializing the optimization matrix
+
+    STATIONARY: Initializes all joint values to the initial value (the current value in the env
+    pci.env->getCurrentJointValues)
+    JOINT_INTERPOLATED: Linearly interpolates between initial value and the joint position specified in InitInfo.data
+    GIVEN_TRAJ: Initializes the matrix to a given trajectory
+
+    In all cases the dt column (if present) is appended the selected method is defined.
+ */
+  enum Type
+  {
+    STATIONARY,
+    JOINT_INTERPOLATED,
+    GIVEN_TRAJ,
+  };
+  /** @brief Specifies the type of initialization to use */
+  Type type;
+  /** @brief Data used during initialization. Use depends on the initialization selected. */
+  trajopt::TrajArray data;
+  /** @brief Default value the final column of the optimization is initialized too if time is being used */
+  double dt = 1.0;
+};
+
+/**
+When cost or constraint element of JSON doc is read, one of these guys gets
+constructed to hold the parameters.
+Then it later gets converted to a Cost object by the hatch method
+*/
+struct TermInfo
+{
+  std::string name;
+  int term_type;
+  int getSupportedTypes() { return supported_term_types_; }
+  //  virtual void fromJson(ProblemConstructionInfo& pci, const Json::Value& v) = 0;
+  virtual void hatch(TrajOptProblem& prob) = 0;
+
+  static TermInfoPtr fromName(const std::string& type);
+
+  /**
+   * Registers a user-defined TermInfo so you can use your own cost
+   * see function RegisterMakers.cpp
+   */
+  typedef TermInfoPtr (*MakerFunc)(void);
+  static void RegisterMaker(const std::string& type, MakerFunc);
+
+  virtual ~TermInfo() = default;
+
+protected:
+  TermInfo(int supported_term_types) : supported_term_types_(supported_term_types) {}
+private:
+  static std::map<std::string, MakerFunc> name2maker;
+  int supported_term_types_;
+};
+
+
+struct ProblemInfo
```

Comment:
Have you tried constructing a very very basic problem, manually, to ensure you understand how this works?

Like a 1dof arm or something?

Reply:
It is actually a good idea. It will help me understand the algorithm more in details. I will do it soon.